subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2327 @@
+
+		LOG_DBG("Found descriptor for format %u, frame %u, MJPEG",
+			format_desc->bFormatIndex, frame_desc->bFrameIndex);
+	} else {
+		struct uvc_format_uncomp_descriptor *format_uncomp_desc = (void *)format_desc;
+
+		fmt->pixelformat = uvc_guid_to_fourcc(format_uncomp_desc->guidFormat);
+
+		LOG_DBG("Found descriptor for format %u, frame %u, GUID '%.4s', pixfmt %04x",
+			format_uncomp_desc->bFormatIndex, frame_desc->bFrameIndex,
+			format_uncomp_desc->guidFormat, fmt->pixelformat);
+	}
+
+	/* Fill the format according to what the host selected */
+	fmt->width = frame_desc->wWidth;
+	fmt->height = frame_desc->wHeight;
+	fmt->pitch = fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE;
+
+	return 0;
+}
+
+static int uvc_get_vs_probe_struct(const struct device *dev, struct uvc_probe *const probe,
+				   const uint8_t request)
+{
+	struct uvc_data *data = dev->data;
+	struct video_format *fmt = &data->video_fmt;
+	int ret;
+
+	ret = uvc_get_vs_probe_format_index(dev, probe, request);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = uvc_get_vs_probe_frame_index(dev, probe, request);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = uvc_get_vs_format_from_desc(dev, fmt);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = uvc_get_vs_probe_frame_interval(dev, probe, request);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = uvc_get_vs_probe_max_size(dev, probe, request);
+	if (ret != 0) {
+		return ret;
+	}
+
+	probe->dwClockFrequency = sys_cpu_to_le32(1);
+	probe->bmFramingInfo = UVC_BMFRAMING_INFO_FID | UVC_BMFRAMING_INFO_EOF;
+	probe->bPreferedVersion = 1;
+	probe->bMinVersion = 1;
+	probe->bMaxVersion = 1;
+	probe->bUsage = 0;
+	probe->bBitDepthLuma = 0;
+	probe->bmSettings = 0;
+	probe->bMaxNumberOfRefFramesPlus1 = 1;
+	probe->bmRateControlModes = 0;
+	probe->bmLayoutPerStream = 0;
+	probe->wKeyFrameRate = sys_cpu_to_le16(0);
+	probe->wPFrameRate = sys_cpu_to_le16(0);
+	probe->wCompQuality = sys_cpu_to_le16(0);
+	probe->wCompWindowSize = sys_cpu_to_le16(0);
+	probe->wDelay = sys_cpu_to_le16(1);
+
+	return 0;
+}
+
+static int uvc_get_vs_probe(const struct device *dev, struct net_buf *const buf,
+			    const struct usb_setup_packet *const setup)
+{
+	struct uvc_data *data = dev->data;
+	size_t size = MIN(sizeof(struct uvc_probe), net_buf_tailroom(buf));
+	struct uvc_probe *probe;
+
+	switch (setup->bRequest) {
+	case UVC_GET_INFO:
+		if (net_buf_tailroom(buf) < 1 || setup->wLength < 1) {
+			return -ENOTSUP;
+		}
+		net_buf_add_u8(buf, UVC_INFO_SUPPORTS_GET);
+		return 0;
+	case UVC_GET_LEN:
+		if (net_buf_tailroom(buf) < 2 || setup->wLength < 2) {
+			return -ENOTSUP;
+		}
+		net_buf_add_le16(buf, sizeof(struct uvc_probe));
+		return 0;
+	case UVC_GET_DEF:
+		net_buf_add_mem(buf, &data->default_probe, size);
+		return 0;
+	case UVC_GET_MIN:
+		__fallthrough;
+	case UVC_GET_RES:
+		__fallthrough;
+	case UVC_GET_MAX:
+		__fallthrough;
+	case UVC_GET_CUR:
+		net_buf_add(buf, size);
+		probe = (void *)buf->data;
+		break;
```

Comment:
This part is a bit confusing, I think it is better to call uvc_get_vs_probe_struct() here and then net_buf_add(), something like

```c
	case UVC_GET_CUR:
		if (size < sizeof(struct uvc_probe)) {
			return -ENOTSUP;
		}

		ret = uvc_get_vs_probe_struct(dev, buf->data, setup->bRequest);
		if (ret == 0) {
			net_buf_add(buf, size);
		}

		return ret;
```